File: examples/typescript/order-repository-dto/order/service.test.ts
import { LineItem, Order } from "./order";
import { InMemoryOrderRepository } from "./repository";
import { FulfillmentService } from "./service";

function assert(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(message);
  }
}

function assertThrows(run: () => void, expectedMessage: string): void {
  try {
    run();
    throw new Error("expected function to throw");
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    assert(message.includes(expectedMessage), `expected error to include "${expectedMessage}", got "${message}"`);
  }
}

const seedItems: LineItem[] = [
  { sku: "kb-100", quantity: 1 },
  { sku: "ms-200", quantity: 2 }
];

const repository = new InMemoryOrderRepository([
  Order.create("ord-ts-4001", "pending", "Ada Lovelace", "Istanbul Warehouse Street 10", seedItems)
]);

const service = new FulfillmentService(repository);
const payload = service.prepareFulfillment("ord-ts-4001");

assert(payload.status === "ready-for-fulfillment", "expected ready-for-fulfillment status");
assert(payload.items.length === 2, "expected two DTO items");
assert(repository.mustGet("ord-ts-4001").status() === "ready-for-fulfillment", "expected saved order status");

assertThrows(() => service.prepareFulfillment("ord-missing"), "was not found");

console.log("order-repository-dto tests passed");

File: examples/go/order-repository-dto/order/service_test.go
package order

import "testing"

func TestPrepareFulfillmentReturnsDTOAndPersistsState(t *testing.T) {
	repository := NewInMemoryRepository([]Order{
		New(
			"ord-4001",
			"pending",
			"Ada Lovelace",
			"Istanbul Warehouse Street 10",
			[]LineItem{
				{SKU: "kb-100", Quantity: 1},
				{SKU: "ms-200", Quantity: 2},
			},
		),
	})

	service := NewFulfillmentService(repository)

	payload, err := service.PrepareFulfillment("ord-4001")
	if err != nil {
		t.Fatalf("prepare fulfillment: %v", err)
	}

	if payload.Status != "ready-for-fulfillment" {
		t.Fatalf("expected ready-for-fulfillment status, got %s", payload.Status)
	}

	if len(payload.Items) != 2 {
		t.Fatalf("expected 2 items, got %d", len(payload.Items))
	}

	saved := repository.MustGet("ord-4001")
	if saved.Status() != "ready-for-fulfillment" {
		t.Fatalf("expected persisted order status to be ready-for-fulfillment, got %s", saved.Status())
	}
}

func TestPrepareFulfillmentRejectsMissingOrder(t *testing.T) {
	repository := NewInMemoryRepository(nil)
	service := NewFulfillmentService(repository)

	if _, err := service.PrepareFulfillment("ord-missing"); err == nil {
		t.Fatal("expected missing order to fail")
	}
}
